### PR TITLE
LlamaPacks: Add ZenGuard llamapack

### PR DESCRIPTION
**Pull request by baur-krykpayev**
# Description

This LlamaPack allows easy integration of ZenGuard AI - the fastest guardrails for GenAI applications. ZenGuard AI protects against 

* Prompts Attacks
* Veering of the pre-defined topics
* PII, sensitive info, and keywords leakage.
* Toxicity
* Etc.

The motivation is that the integration with LlamaIndex was one of the most commonly asked features.

## New Package?

Did I fill in the `tool.llamahub` section in the `pyproject.toml` and provide a detailed README.md for my new integration or package?

- [x] Yes

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [x] Yes

## Type of Change

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Added new unit/integration tests
- [x] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

## Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added Google Colab support for the newly added notebooks.
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama-index-packs/llama-index-packs-zenguard/poetry.lock
Comment (logan-markewich) [inline llama-index-packs/llama-index-packs-zenguard/poetry.lock:1]: no need to commit this file

File: llama-index-packs/llama-index-packs-zenguard/pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+disallow_untyped_defs = true
+# Remove venv skip when integrated with pre-commit
+exclude = ["_static", "build", "examples", "notebooks", "venv"]
+ignore_missing_imports = true
+python_version = "3.8"
+
+[tool.poetry]
+authors = ["ZenGuard Team <hello@zenguard.ai>"]
+description = "llama-index packs zenguard guardrails integration"
+license = "MIT"
```
Comment (nerdai) [inline llama-index-packs/llama-index-packs-zenguard/pyproject.toml:29]: please add below this line the following: `exclude = ["**/BUILD"]`

**Review by nerdai — APPROVED** [review]
I think this is a good star @baur-krykpayev! Thanks for the PR. I am curious to know how the users who asked for this integration wanted it to look like? This is a light wrapper as you indicated over `ZenGuard`. I am wondering if it make sense to in the future to create an "OutputParser" extension or perhaps some other appropriate extension into the LlamaIndex ecosystem.

**Review by nerdai — COMMENTED** [review]
@baur-krykpayev do you mind adding me as a contributor to your the ZenGuard fork? I am trying to fix tests... Alternatively, you can do it yourself but you just need to install `pants` and run a couple of commands.

**Comment by baur-krykpayev** [comment]
> @baur-krykpayev do you mind adding me as a contributor to your the ZenGuard fork? I am trying to fix tests... Alternatively, you can do it yourself but you just need to install `pants` and run a couple of commands.

Thank you for helping out @nerdai! I really appreciate it. I have invited you to the fork https://github.com/ZenGuard-AI/llama_index

**Comment by nerdai** [comment]
> > @baur-krykpayev do you mind adding me as a contributor to your the ZenGuard fork? I am trying to fix tests... Alternatively, you can do it yourself but you just need to install `pants` and run a couple of commands.
> 
> Thank you for helping out @nerdai! I really appreciate it. I have invited you to the fork https://github.com/ZenGuard-AI/llama_index

Thanks! I have added the BUILD files that pants needs. Now tests should actually run. Fingers crossed 🤞that they pass!

File: llama-index-packs/llama-index-packs-zenguard/pyproject.toml (hunk 1 above)
Comment (nerdai) [reply llama-index-packs/llama-index-packs-zenguard/pyproject.toml:30]: don't worry about this one -- i took care of it :)

**Comment by baur-krykpayev** [comment]
> I think this is a good star @baur-krykpayev! Thanks for the PR. I am curious to know how the users who asked for this integration wanted it to look like? This is a light wrapper as you indicated over `ZenGuard`. I am wondering if it make sense to in the future to create an "OutputParser" extension or perhaps some other appropriate extension into the LlamaIndex ecosystem.

Ideally, we would like to have both InputParser and OutputParser enabled. However, we are currently a bit short on both time and manpower. Hence, we would like to start from the packs and check how it lands with the users.

**Comment by nerdai** [comment]
> > I think this is a good star @baur-krykpayev! Thanks for the PR. I am curious to know how the users who asked for this integration wanted it to look like? This is a light wrapper as you indicated over `ZenGuard`. I am wondering if it make sense to in the future to create an "OutputParser" extension or perhaps some other appropriate extension into the LlamaIndex ecosystem.
> 
> Ideally, we would like to have both InputParser and OutputParser enabled. However, we are currently a bit short on both time and manpower. Hence, we would like to start from the packs and check how it lands with the users.

cool... yeah the bandwidth stuff always an issue. Your approach makes a lot of sense to me!

**Comment by nerdai** [comment]
@baur-krykpayev looks unit tests are failing because of a missing api key. Can we provide a mock key here instead?

File: llama-index-packs/llama-index-packs-zenguard/poetry.lock
Comment (baur-krykpayev) [reply llama-index-packs/llama-index-packs-zenguard/poetry.lock:1]: removed

**Comment by baur-krykpayev** [comment]
> @baur-krykpayev looks like unit tests are failing because of a missing api key. Can we provide a mock key here instead?

hard-coded a mock key that is going to be whitelisted only for LlamaIndex tests. Hopefully this will fix the tests.

**Comment by baur-krykpayev** [comment]
ZenGuard AI client uses python 3.9. Hence the Unit Testing for 3.8 is failing. @nerdai what is a course of action in this case?

**Comment by nerdai** [comment]
> ZenGuard AI client uses python 3.9. Hence the Unit Testing for 3.8 is failing. @nerdai what is a course of action in this case?

Yeah, looks like we need to set the constraints in our pants BUILD files to only run with certain python versions — skipping python 3.8 in this case.

**Comment by baur-krykpayev** [comment]
Thank you for the speedy review!

**Pull request merged**